src/System.Memory/src/System/Span.cs

```diff
@@ -0,0 +1,401 @@
+            _pinnable = null;
+            _byteOffset = new IntPtr(pointer);
+        }
+
+        /// <summary>
+        /// Create a new span over a portion of a regular managed object. This can be useful
+        /// if part of a managed object represents a "fixed array." This is dangerous because 
+        /// "length" is not checked, nor is the fact that "rawPointer" actually lies within the object.
+        /// </summary>
+        /// <param name="obj">The managed object that contains the data to span over.</param>
+        /// <param name="objectData">A reference to data within that object.</param>
+        /// <param name="length">The number of <typeparamref name="T"/> elements the memory contains.</param>
+        /// <exception cref="System.ArgumentNullException">
+        /// Thrown when the specified object is null.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="length"/> is negative.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Span<T> DangerousCreate(object obj, ref T objectData, int length)
+        {
+            if (obj == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.obj);
+            if (length < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.length);
+
+            Pinnable<T> pinnable = Unsafe.As<Pinnable<T>>(obj);
+            IntPtr byteOffset = Unsafe.ByteOffset<T>(ref pinnable.Data, ref objectData);
+            return new Span<T>(pinnable, byteOffset, length);
+        }
+
+        // Constructor for internal use only.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private Span(Pinnable<T> pinnable, IntPtr byteOffset, int length)
+        {
+            Debug.Assert(length >= 0);
+
+            _length = length;
+            _pinnable = pinnable;
+            _byteOffset = byteOffset;
+        }
+
+        /// <summary>
+        /// The number of items in the span.
+        /// </summary>
+        public int Length => _length;
+
+        /// <summary>
+        /// Returns true if Length is 0. 
+        /// </summary>
+        public bool IsEmpty => _length == 0;
+
+        /// <summary>
+        /// Returns a reference to specified element of the Span.
+        /// </summary>
+        /// <param name="index"></param>
+        /// <returns></returns>
+        /// <exception cref="System.IndexOutOfRangeException">
+        /// Thrown when index less than 0 or index greater than or equal to Length
+        /// </exception>
+        public ref T this[int index]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                if ((uint)index >= ((uint)_length))
+                    ThrowHelper.ThrowIndexOutOfRangeException();
+
+                IntPtr adjustedByteOffset = _byteOffset.Add<T>(index);
```

Comment:
Adding the index via `Unsafe.Add<T>(ref, index)` is likely going to produce better code. `Unsafe. Add<T>(ref, index)` is a leaf function with just a few IL instructions. `SpanHelpers.Add` has branches and calls of other functions - not sure whether the JIT is going to be able to optimize it out completely.